design-history/exploration-reports/2019.06-unixfsv2-spike-01.md

```diff
@@ -0,0 +1,296 @@
+	gid Int # make another schema which deliberately excludes this.  try to match this one (the one with more info) first.
+}
+
+# You should be able to imagine sending a fancy graphsync request with some structure like:
+#
+#  {"schema": {{schemaCIDofUnixfs}}, "selector": {{selectorSpec}},
+#   "schemaPatch": {"./dir/transparently/through/to/file": {{schemaCIDofRabinLayout}},
+#   "startAt": {{rootBlockCID+path}} }
+#
+# This kind of fancy graphsync request can be used to make a query which does "the right thing" with
+# all advanced layouts transparently; then when reaching the 'patch' area, perhaps the selector spec
+# wants to use the seeded-psuedorandomness feature, so it can issue queries like this across many peers.
+#
+# Perhaps we won't actually need the 'patch' feature, because in practice if doing queries with lots
+# of fancy sharded requests, prooobably you'll do the path *up to* that first, in a separate query,
+# But either way, you can see how using schemas for signaling is useful here.
+```
+
+Interesting Highlights
+----------------------
+
+### Using various schemas as a feature!
+
+Observe the comments on the `Attribs.uid` and `Attribs.gid` fields.
+
+By using several different (but significantly overlapping) schemas to describe variations
+in which attributes we're storing, we can make unixfsv2 into an extensible yet still
+very well defined system.
+
+Concretely: If the `uid` and `gid` elements are found present in the first `Attribs` map
+processed in a tree of data, then we match the first schema (the one with more info),
+and we can now expect *all future* `Attribs` maps to consistently include the same data.
+If the `uid` or `gid` elements are found to be missing, we would flunk out of matching
+the first schema, and try a second schema which does not include them... and again,
+now expect *all future* `Attribs` maps to consistently *not* include these data.
+
+This kind of deep consistency makes it easier to build applications,
+especially applications which may want to change their behavior
+or apply distinct transformations depending on which of the schemas matches.
+
+### Advanced layout usage sites are marked with angle brackets
+
+E.g., on the line `members {Filename:DirEnt}<HAMT>`, it's the `<` and `>`
+characters which denote that we'll be using an advanced layout here.
+
+We might also consider using different characters for this denotation.
+(Angle brackets are reminiscent of generics in Java syntax;
+this may or may not be a helpful reminiscence.)
+
+We might also consider a different approach entirely: ditching the ability to
+declare advanced layouts on inline type terms at all, and instead using
+something more like the `representation` keyword.
+Such an approach would syntactically require a fully named type for any
+structure that wants to use an advanced layout.
+Given that the most common applications of advanced layouts are for map, list,
```

Comment:
If we have an “in band” mechanism for attaching schemas to composites, do we actually *need* to do all of this definition up front?

For example, if we have a definition of “Bytesish” that is something along the lines of “implements `read()` and `length()`” we would use that in this base schema. The schema for `RabinBytes` could then live in the composite and we don’t need to increase the volume of syntax and definitions here. This is sort of nice as it lets you provide general assurances over a broad graph and then much more granular assertions as you work your way down the graph.

If someone wants to be very explicit and say “this must be RabinBytes” we obviously have the ability to do that, but it would be much nicer to just say “I need a thing that I can treat as binary” and leave the door open to future Bytes implementations. As long as we attach the schema to the composite you’ll still have codegen and stricter validation once you have one of these nodes.